### PR TITLE
move Longhorn storageclass management into setting controller to prevent upgrade failure due to changed storageclass

### PR DESCRIPTION
**Pull request by PhanLe1010**
We move the default Longhorn storageclass into setting controller so that it is not managed by Helm anymore. This helps to prevent the failure while upgrading Longhorn using Helm and the Longhorn default storageclass  is accidentally modified by users.

Now, the storageclass ConfigMap will be the source of true. Any changes to this ConfigMap will result in creation of new longhorn storageclass.

longhorn/longhorn#1527

### CODE REVIEW

File: controller/setting_controller.go [hunk 1]
```diff
@@ -403,6 +428,90 @@ func (sc *SettingController) updatePriorityClass() error {
...
+
+	storageclass, err = sc.ds.CreateStorageClass(storageclass)
+	if err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			return nil
+		}
+		return err
+	}
+
+	sc.logger.Infof("Updated Longhorn storagecclass: %v", storageclass)
```
Comment (yasker) [inline controller/setting_controller.go:472]: typo. `StorageClass`.

File: deploy/uninstall/uninstall.yaml [hunk 2]
```diff
@@ -64,7 +64,7 @@ spec:
     spec:
       containers:
       - name: longhorn-uninstall
-        image: longhornio/longhorn-manager:master
+        image: phanle1010/longhorn-manager:sc-12
```
Comment (yasker) [inline deploy/uninstall/uninstall.yaml:67]: ?

File: controller/setting_controller.go [hunk 3]
```diff
@@ -403,6 +428,90 @@ func (sc *SettingController) updatePriorityClass() error {
...
+				return nil
+			}
+			return nil
+		}
+		return err
+	}
+
+	storageclass, err := buildStorageClassManifestFromConfigMap(storageConfigMap)
+	if err != nil {
+		return err
```
Comment (yasker) [inline controller/setting_controller.go:447]: So a failure to parse the config map will stop the setting controller from reconciliation?

Also, the extra setting name is weird. Now I think it's better to have `ConfigMapController` to take in the change instead.

**Review by yasker — CHANGES_REQUESTED** [review]

File: controller/setting_controller.go (hunk 3 above)
Comment (PhanLe1010) [reply controller/setting_controller.go:447]: Yes, I agree that the extra setting is a bit weird.

File: controller/setting_controller.go [hunk 4]
```diff
@@ -403,6 +428,90 @@ func (sc *SettingController) updatePriorityClass() error {
...
 }
 
+func (sc *SettingController) updateDefaultStorageClass() error {
+	storageConfigMap, err := sc.ds.GetConfigMapRO(types.DefaultStorageClassConfigMapName)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			// The storageclass configmap is deleted therefore we delete the actual storageclass
+			err := sc.ds.DeleteStorageClass(types.DefaultStorageClassName)
+			if err != nil && !datastore.ErrorIsNotFound(err) {
+				return nil
```
Comment (c3y1huang) [inline controller/setting_controller.go:438]: Better to return the error instead?

**Comment by PhanLe1010** [comment]
@c3y1huang 

Since we decided to to create a new controller to handle storageclass configmap. I have closed this PR and create a new one at https://github.com/longhorn/longhorn-manager/pull/738

Thank you for your review! Can we discuss about your review at https://github.com/longhorn/longhorn-manager/pull/738 instead?

**Comment by PhanLe1010** [comment]
Close this PR because creating a replacement one at https://github.com/longhorn/longhorn-manager/pull/738

File: controller/setting_controller.go (hunk 4 above)
Comment (PhanLe1010) [reply controller/setting_controller.go:438]: Yeah. We should return err here. Thank you! Fixed at https://github.com/longhorn/longhorn-manager/pull/738

File: controller/setting_controller.go [hunk 5]
```diff
@@ -403,6 +428,90 @@ func (sc *SettingController) updatePriorityClass() error {
...
+
+	sc.logger.Infof("Updated Longhorn storagecclass: %v", storageclass)
+
+	return nil
+}
+
+func buildStorageClassManifestFromConfigMap(cfm *v1.ConfigMap) (*storagev1.StorageClass, error) {
+	data, ok := cfm.Data["storageclass.yaml"]
+	if !ok {
+		return nil, fmt.Errorf("cannot find storageclass.yaml inside the default storageclass configmap")
```
[the remark this answers is not in the document]
Comment (PhanLe1010) [reply controller/setting_controller.go:480]: It doesn't seems to have a func `fmt.Error`

File: datastore/kubernetes.go [hunk 6]
```diff
@@ -176,6 +177,23 @@ func (s *DataStore) DeletePod(name string) error {
...
+func (s *DataStore) GetStorageClassRO(scName string) (*storagev1.StorageClass, error) {
+	return s.storageclassLister.Get(scName)
+}
+
+// DeleteStorageClass deletes StorageClass with the given name
+func (s *DataStore) DeleteStorageClass(scName string) error {
+	return s.kubeClient.StorageV1().StorageClasses().Delete(scName, nil)
+}
+
+// CreateStorageClass creates StorageClass with the given name
```
[the remark this answers is not in the document]
Comment (PhanLe1010) [reply datastore/kubernetes.go:192]: Thank you! Fixed

**Pull request closed without merging**